### ISSUE TITLE
[BUG] corrupted packages with apktool for sdk >= 30 (Android 10) (aapt/aapt2)

### ISSUE THREAD

**Issue by eladkarako** [iBotPeaches/Apktool#2769]
for some quite time now I'm getting corrupted packages,  
is apktool still functional?
I'm currently on build `2.6.2-adc942-SNAPSHOT`, and just extracting and building renders an invalid (non-installable) apk.

I assume the cause is parsing res and resources.arsc since the result apk(s) have broken icon, and the packages reported as uninstallable ('problem while parsing the package').
I'm properly signing and aligning.

it seems that even the simplest apk are broken regardless if they are built using the aapt2 or not.

`apktool.yml` set to sdk 29 renders an installable apk (with visible icon), but most of the apps that requires file access on Android 11+ won't work properly with that sdk. so I effectively have no way of working with apktool :..[

<hr/>

Ps. are there requirement for java version, 
and is JDK 17.0.2 (currently latest) requires any `--add-opens` or `--add-exports` ?

currently using oracle jdk1.8.0_151

**Comment by iBotPeaches** [comment]
> is apktool still functional?

Yes

> uninstallable ('problem while parsing the package').

The logcat would have the exact reason it was rejected.

> apktool.yml set to sdk 29 renders an installable apk,

Saying this suggests sdk28 works perfectly which would be something to investigate, is that true?


> Ps. are there requirement for java version,

8+ 

> and is JDK 17.0.2 (currently latest) requires any --add-opens or --add-exports ?

I'm not positive what those means to comment on that.

**Comment by eladkarako** [comment]
thank you for answering the above issues.
after addressing some known issue: https://gist.github.com/eladkarako/76ace7879872b9cc9b4b5817e08392f4  
I was able to build with either aapt(1) and aapt2 successfully,  
I've also switched to oracle JDK 17, which does not seems to matter much.
I've tried leaving sdk 31, changing to sdk 30 and sdk 29.
only when I have switched to sdk 29 I could see an icon in the apk and install it.
sdk 31/30 was not installable no matter what I did.

I understand that there is no much information to go on in this bug, 
so going to close it for now. as far as I can tell, trying to modify apks that were compiled to (no matter the minimum sdk) sdk 31(+) will no re-build properly. I suspect the parsing of the res + resources.arsc.

I haven't found any suitable alternative to apktool.

**Comment by eladkarako** [comment]
`jarsigner` and `apksigner` (from build tools) didn't worked for me,  
but `uber-apk-signer-1.2.1.jar` generated a working apk from what was found in the `dist` folder.

afterwards, I was unpacking the apk for copying some stuff and zipping it back with either 7zip/zip/jar,  
and signing (with `uber-apk-signer`) the result was an invalid apk.
for that I've used gnu zip, and zero compression and not storing folder entries: `zip -0Dr repacked_dist.apk .`  following with signing (with `uber-apk-signer`),  

the result was a working apk.

not sure way but from SDK 30,  
a too strong zip compression seems to break the apk.
maybe the validator is having a hard time reading the constructed files for some reason,  
and the zero-compression solved it. it seems like a workaround for a bug (possibly) in the packing.

I've also noticed the signer uses zipalign's `-p` argument for `page align stored shared object files`.

there were a lot of other changes, with the keystore creation-parameters, JDK, and build-tools, but the zip compression seems to be the one working around whatever broke before..
